src/Native/System.Native/pal_networkchange.cpp

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include "pal_config.h"
+
+#if HAVE_LINUX_RTNETLINK_H
+#include "pal_types.h"
+#include "pal_networkchange.h"
+#include "pal_utilities.h"
+
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/uio.h>
+#include <sys/socket.h>
+#include <net/if.h>
+#include <linux/netlink.h>
+#include <linux/rtnetlink.h>
+
+#pragma clang diagnostic ignored "-Wcast-align"
+
+extern "C" int CreateNetworkChangeListenerSocket()
+{
+    struct sockaddr_nl sa = { };
+    sa.nl_family = AF_NETLINK;
+    sa.nl_groups = RTMGRP_LINK | RTMGRP_IPV4_IFADDR;
+    int sock = socket(AF_NETLINK, SOCK_RAW, NETLINK_ROUTE);
+    bind(sock, reinterpret_cast<sockaddr*>(&sa), sizeof(sa));
+    return sock;
+}
+
+extern "C" int32_t CloseNetworkChangeListenerSocket(int socket)
+{
+    int result = close(socket);
+    return result;
+}
+
+extern "C" NetworkChangeKind ReadSingleEvent(int sock)
+{
+    char buffer[4096];
```

Comment:
It's not very clearly documented, but this is the size that is used in the example on the manpage (netlink 7), and I've never seen it be too small.